src/compile.rs

```diff
@@ -329,6 +346,118 @@ pub fn build_func_map<'a>(
     Ok(decl_map)
 }
 
+pub fn validate_fs_context_duplicates(
+    fsc_rules: BTreeMap<String, BTreeSet<&FileSystemContextRule>>,
+) -> Result<(), CascadeErrors> {
+    let mut errors = CascadeErrors::new();
+
+    'key_loop: for v in fsc_rules.values() {
+        // We only have 1 or 0 elements, thus we cannot have a semi duplicate
+        if v.len() <= 1 {
+            continue;
+        }
+        let mut error: Option<CompileError> = None;
+        for rule in v {
+            match rule.fscontext_type {
+                // If we ever see a duplicate of xattr task or trans we know something is wrong
+                FSContextType::XAttr | FSContextType::Task | FSContextType::Trans => {
+                    error = Some(add_or_create_compile_error(error,
+                        "Duplicate filesystem context.",
+                        &rule.file,
+                        unwrap_or_return!(rule.fs_name.get_range(), Err(CascadeErrors::from(InternalError::new()))),
+                        &format!("Found multiple different filesystem type declarations for filesystem: {}", rule.fs_name)));
+                }
+                FSContextType::GenFSCon => {
+                    // genfscon gets more complicated.  We can have similar rules as long as the paths are different.
+                    // If we find a genfscon with the same path, they must have the same context and object type.
+                    if let Some(path) = &rule.path {
+                        // Look through the rules again
+                        for inner_rule in v {
+                            // Only check path if it was provided as part of the rule
+                            if let Some(inner_path) = &inner_rule.path {
+                                // If our paths match, check if our contexts match
+                                if path == inner_path && rule.context != inner_rule.context {
+                                    // error is not None so we have already found something, so we just
+                                    // need to add a new error message
+                                    if let Some(unwrapped_error) = error {
+                                        error = Some(unwrapped_error.add_additional_message(&inner_rule.file,
```

Comment:
This block looks identical to the similar one below except for the range.  I feel like defining a local closure that takes a range and then calling it here will be cleaner.

Reply:
Alternatively, why can't we just reuse add_or_create_compile_error() again?

Reply:
In the else we have a weird condition where we need to create the error then add a new message to it immediately.